### PR TITLE
ES - Starting missing shard tasks

### PR DESCRIPTION
**Pull request by tbhanu-amzn**
### Description
ES -  Starting missing shard tasks
 
### Issues Resolved
ES -  Starting missing shard tasks
 
### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt [hunk 1]
```diff
@@ -794,6 +806,18 @@ class IndexReplicationTask(id: Long, type: String, action: String, description:
             ShardReplicationExecutor.TASK_NAME, replicationParams)
     }
 
+    override fun onIndexShardClosed(shardId: ShardId, indexShard: IndexShard?, indexSettings: Settings) {
+    }
```
Comment (krishna-ggk) [inline src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt:810]: A comment would be useful as to why we are keeping this no-op.

File: src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt [hunk 2]
```diff
@@ -794,6 +806,18 @@ class IndexReplicationTask(id: Long, type: String, action: String, description:
...
+
+    override fun onIndexRemoved(indexService: IndexService,
+                                reason: IndicesClusterStateService.AllocatedIndices.IndexRemovalReason) {
+        // cancel the index task only if the index is closed
+        val indexMetadata = indexService.indexSettings.indexMetadata
+        if(indexMetadata.state != IndexMetadata.State.OPEN) {
```
Comment (krishna-ggk) [inline src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt:816]: When else is indexRemoved triggered?

File: src/test/kotlin/com/amazon/elasticsearch/replication/integ/rest/ClusterRerouteLeaderIT.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.junit.Assert
+import org.junit.Ignore
+import java.util.concurrent.TimeUnit
+
+@MultiClusterAnnotations.ClusterConfigurations(
+        MultiClusterAnnotations.ClusterConfiguration(clusterName = LEADER),
+        MultiClusterAnnotations.ClusterConfiguration(clusterName = FOLLOWER)
+)
+
+class ClusterRerouteLeaderIT : MultiClusterRestTestCase() {
```
Comment (krishna-ggk) [inline src/test/kotlin/com/amazon/elasticsearch/replication/integ/rest/ClusterRerouteLeaderIT.kt:25]: Nit - seems like there is potential to share code (esp changeTemplate etc methods).

Can we combine them in one class may be OR extract the common code?

**Review by krishna-ggk — COMMENTED** [review]
LGTM - just couple questions

File: src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt (hunk 1 above)
Comment (tbhanu-amzn) [reply src/main/kotlin/com/amazon/elasticsearch/replication/task/index/IndexReplicationTask.kt:810]: fixed by Sai

File: src/test/kotlin/com/amazon/elasticsearch/replication/integ/rest/ClusterRerouteLeaderIT.kt (hunk 3 above)
Comment (tbhanu-amzn) [reply src/test/kotlin/com/amazon/elasticsearch/replication/integ/rest/ClusterRerouteLeaderIT.kt:25]: extracting common code to the base class to keep the code separate for leader and follower

**Review by krishna-ggk — APPROVED** [review]

**Pull request merged**